File: phase10/cards.py
"""Card model and deck construction.

Stock deck is 108 cards: numbers 1-12 in four colors, two copies each (96),
plus 8 Wilds and 4 Skips. Deck composition is configurable because the AP
items "Wild Card" and "Skip Card" literally add cards to the draw pile.
"""

from __future__ import annotations

import random
from dataclasses import dataclass
from enum import Enum

MIN_RANK = 1
MAX_RANK = 12
COPIES_PER_RANK = 2

STOCK_WILDS = 8
STOCK_SKIPS = 4


class Color(Enum):
    RED = "red"
    BLUE = "blue"
    GREEN = "green"
    YELLOW = "yellow"


class Kind(Enum):
    NUMBER = "number"
    WILD = "wild"
    SKIP = "skip"


@dataclass(frozen=True, order=True)
class Card:
    kind: Kind
    rank: int | None = None
    color: Color | None = None

    @property
    def is_number(self) -> bool:
        return self.kind is Kind.NUMBER

    @property
    def is_wild(self) -> bool:
        return self.kind is Kind.WILD

    @property
    def is_skip(self) -> bool:
        return self.kind is Kind.SKIP

    @property
    def points(self) -> int:
        """Official scoring: 1-9 = 5, 10-12 = 10, Skip = 15, Wild = 25."""
        if self.kind is Kind.WILD:
            return 25
        if self.kind is Kind.SKIP:
            return 15
        return 5 if self.rank <= 9 else 10

    def __str__(self) -> str:
        if self.kind is Kind.WILD:
            return "W"
        if self.kind is Kind.SKIP:
            return "S"
        return f"{self.rank}{self.color.value[0].upper()}"


WILD = Card(Kind.WILD)
SKIP = Card(Kind.SKIP)


def number_card(rank: int, color: Color) -> Card:
    if not MIN_RANK <= rank <= MAX_RANK:
        raise ValueError(f"rank {rank} out of range")
    return Card(Kind.NUMBER, rank, color)


def build_deck(wilds: int = STOCK_WILDS, skips: int = STOCK_SKIPS) -> list[Card]:
    """Build a deck. `wilds` and `skips` are driven by AP item counts."""
    if not 0 <= wilds <= STOCK_WILDS:
        raise ValueError(f"wilds must be 0..{STOCK_WILDS}, got {wilds}")
    if not 0 <= skips <= STOCK_SKIPS:
        raise ValueError(f"skips must be 0..{STOCK_SKIPS}, got {skips}")

    deck: list[Card] = []
    for rank in range(MIN_RANK, MAX_RANK + 1):
        for color in Color:
            deck.extend([number_card(rank, color)] * COPIES_PER_RANK)
    deck.extend([WILD] * wilds)
    deck.extend([SKIP] * skips)
    return deck


def shuffled_deck(rng: random.Random, wilds: int = STOCK_WILDS, skips: int = STOCK_SKIPS) -> list[Card]:
    deck = build_deck(wilds, skips)
    rng.shuffle(deck)
    return deck


def hand_score(cards: list[Card]) -> int:
    return sum(c.points for c in cards)

File: docs/src/cards.js
/**
 * Card model and deck construction.
 *
 * A direct port of phase10/game/cards.py. The Python engine stays the
 * reference implementation -- tools/export_fixtures.py emits its verdicts and
 * test/crosscheck.mjs proves this port agrees with them.
 *
 * Stock deck is 108 cards: numbers 1-12 in four colors, two copies each (96),
 * plus 8 Wilds and 4 Skips. Deck composition is configurable because the AP
 * items "Wild Card" and "Skip Card" literally add cards to the draw pile.
 */

export const MIN_RANK = 1;
export const MAX_RANK = 12;
export const COPIES_PER_RANK = 2;

export const STOCK_WILDS = 8;
export const STOCK_SKIPS = 4;

/** Declaration order matters: _solveColor returns the first colour that fits. */
export const COLORS = ["red", "blue", "green", "yellow"];

export const KIND = {
  NUMBER: "number",
  WILD: "wild",
  SKIP: "skip",
};

export function numberCard(rank, color) {
  if (rank < MIN_RANK || rank > MAX_RANK) {
    throw new RangeError(`rank ${rank} out of range`);
  }
  if (!COLORS.includes(color)) {
    throw new RangeError(`unknown color ${color}`);
  }
  return { kind: KIND.NUMBER, rank, color };
}

export const WILD = Object.freeze({ kind: KIND.WILD, rank: null, color: null });
export const SKIP = Object.freeze({ kind: KIND.SKIP, rank: null, color: null });

export const isNumber = (c) => c.kind === KIND.NUMBER;
export const isWild = (c) => c.kind === KIND.WILD;
export const isSkip = (c) => c.kind === KIND.SKIP;

/** Official scoring: 1-9 = 5, 10-12 = 10, Skip = 15, Wild = 25. */
export function points(card) {
  if (card.kind === KIND.WILD) return 25;
  if (card.kind === KIND.SKIP) return 15;
  return card.rank <= 9 ? 5 : 10;
}

export function handScore(cards) {
  return cards.reduce((total, c) => total + points(c), 0);
}

/** Matches Card.__str__ in the Python model, so traces line up. */
export function cardToString(card) {
  if (card.kind === KIND.WILD) return "W";
  if (card.kind === KIND.SKIP) return "S";
  return `${card.rank}${card.color[0].toUpperCase()}`;
}

/** Inverse of cardToString -- used to read fixtures emitted by Python. */
export function cardFromString(s) {
  if (s === "W") return { ...WILD };
  if (s === "S") return { ...SKIP };
  const m = /^(\d{1,2})([RBGY])$/.exec(s);
  if (!m) throw new Error(`cannot parse card ${JSON.stringify(s)}`);
  const color = COLORS.find((c) => c[0].toUpperCase() === m[2]);
  return numberCard(Number(m[1]), color);
}

/** Filename of a card's rendered face. Mirrors cards.card_filename. */
export function cardFilename(card) {
  if (card.kind === KIND.WILD) return "wild.png";
  if (card.kind === KIND.SKIP) return "skip.png";
  return `${card.color}_${String(card.rank).padStart(2, "0")}.png`;
}

export function buildDeck(wilds = STOCK_WILDS, skips = STOCK_SKIPS) {
  if (wilds < 0 || wilds > STOCK_WILDS) {
    throw new RangeError(`wilds must be 0..${STOCK_WILDS}, got ${wilds}`);
  }
  if (skips < 0 || skips > STOCK_SKIPS) {
    throw new RangeError(`skips must be 0..${STOCK_SKIPS}, got ${skips}`);
  }
  const deck = [];
  for (let rank = MIN_RANK; rank <= MAX_RANK; rank++) {
    for (const color of COLORS) {
      for (let i = 0; i < COPIES_PER_RANK; i++) deck.push(numberCard(rank, color));
    }
  }
  for (let i = 0; i < wilds; i++) deck.push({ ...WILD });
  for (let i = 0; i < skips; i++) deck.push({ ...SKIP });
  return deck;
}
